File: ios/Packages/HapiKit/Sources/HapiProtocol/Models/SessionPatch.swift
import Foundation

/// A `{version, value}` wrapper carried by structured `session-updated`
/// patches for `metadata` / `agentState` / `todos` / `teamState`.
///
/// The version is the only safe way to reject stale patches: the dual SSE
/// connections have no shared ordering, so `value` must be applied — and the
/// matching watermark stored — only when `version` is strictly greater than
/// the cached one (see `isNewerVersionedPatch(patchVersion:cachedVersion:)`).
/// Never assign the wrapper itself into a `Session`.
public struct VersionedValue<Value: Codable & Equatable & Sendable>: Codable, Equatable, Sendable {
    public var version: Int
    /// `nil` covers wire-`null` (explicit clear, e.g. team deletion).
    public var value: Value?

    public init(version: Int, value: Value?) {
        self.version = version
        self.value = value
    }
}

/// A patch field whose wire type is `T | null`, where "key absent" and
/// "key present with null" mean different things.
///
/// `SessionPatch.model = nil` means the patch does not touch `model`;
/// `.null` means the patch explicitly clears it. Swift's synthesized Codable
/// (`decodeIfPresent`) conflates the two, so `SessionPatch` decodes these by
/// hand via `contains` + `decodeNil`.
public enum PatchField<Value: Codable & Equatable & Sendable>: Equatable, Sendable {
    case null
    case value(Value)

    /// The value as it would appear after applying: `nil` for `.null`.
    public var wireValue: Value? {
        if case .value(let value) = self { return value }
        return nil
    }
}

/// A structured `session-updated` patch.
///
/// Mirrors `SessionPatchSchema` (`shared/src/schemas.ts:371-404`), which is
/// zod `.strict()`: unknown keys make decoding throw. That strictness is
/// load-bearing — `SessionUpdatedData` discriminates the
/// `Session | SessionPatch` union by attempting `Session` first and falling
/// back to the strict patch, so a malformed full session must not silently
/// half-decode as a lenient patch.
public struct SessionPatch: Equatable, Sendable {
    public var active: Bool?
    public var thinking: Bool?
    /// Carried on the wire but NOT applied by `applySessionDetailPatch`
    /// (mirroring `web/src/lib/sessionPatch.ts`).
    public var activeTurnStartedAt: PatchField<Int>?
    public var activeAt: Int?
    public var updatedAt: Int?
    public var metadata: VersionedValue<SessionMetadata>?
    public var agentState: VersionedValue<AgentState>?
    public var todos: VersionedValue<[TodoItem]>?
    public var teamState: VersionedValue<JSONValue>?
    public var model: PatchField<String>?
    public var modelReasoningEffort: PatchField<String>?
    public var effort: PatchField<String>?
    public var serviceTier: PatchField<String>?
    public var permissionMode: PermissionMode?
    public var collaborationMode: CodexCollaborationMode?
    public var copilotAgentMode: CopilotAgentMode?
    public var backgroundTaskCount: Int?
    /// Bare refetch trigger for `GET /api/sessions/:id/scratchlist`; never
    /// applied onto the `Session` (which carries no such field).
    public var scratchlistUpdatedAt: Int?

    public init(
        active: Bool? = nil,
        thinking: Bool? = nil,
        activeTurnStartedAt: PatchField<Int>? = nil,
        activeAt: Int? = nil,
        updatedAt: Int? = nil,
        metadata: VersionedValue<SessionMetadata>? = nil,
        agentState: VersionedValue<AgentState>? = nil,
        todos: VersionedValue<[TodoItem]>? = nil,
        teamState: VersionedValue<JSONValue>? = nil,
        model: PatchField<String>? = nil,
        modelReasoningEffort: PatchField<String>? = nil,
        effort: PatchField<String>? = nil,
        serviceTier: PatchField<String>? = nil,
        permissionMode: PermissionMode? = nil,
        collaborationMode: CodexCollaborationMode? = nil,
        copilotAgentMode: CopilotAgentMode? = nil,
        backgroundTaskCount: Int? = nil,
        scratchlistUpdatedAt: Int? = nil
    ) {
        self.active = active
        self.thinking = thinking
        self.activeTurnStartedAt = activeTurnStartedAt
        self.activeAt = activeAt
        self.updatedAt = updatedAt
        self.metadata = metadata
        self.agentState = agentState
        self.todos = todos
        self.teamState = teamState
        self.model = model
        self.modelReasoningEffort = modelReasoningEffort
        self.effort = effort
        self.serviceTier = serviceTier
        self.permissionMode = permissionMode
        self.collaborationMode = collaborationMode
        self.copilotAgentMode = copilotAgentMode
        self.backgroundTaskCount = backgroundTaskCount
        self.scratchlistUpdatedAt = scratchlistUpdatedAt
    }
}

extension SessionPatch: Codable {
    private enum CodingKeys: String, CodingKey, CaseIterable {
        case active
        case thinking
        case activeTurnStartedAt
        case activeAt
        case updatedAt
        case metadata
        case agentState
        case todos
        case teamState
        case model
        case modelReasoningEffort
        case effort
        case serviceTier
        case permissionMode
        case collaborationMode
        case copilotAgentMode
        case backgroundTaskCount
        case scratchlistUpdatedAt
    }

    public init(from decoder: Decoder) throws {
        // Mirror zod `.strict()`: any unknown key rejects the whole patch.
        try rejectUnknownKeys(
            in: decoder,
            known: Set(CodingKeys.allCases.map(\.stringValue)),
            payloadName: "SessionPatch"
        )
        let container = try decoder.container(keyedBy: CodingKeys.self)
        active = try container.decodeIfPresent(Bool.self, forKey: .active)
        thinking = try container.decodeIfPresent(Bool.self, forKey: .thinking)
        activeTurnStartedAt = try Self.decodePatchField(Int.self, in: container, forKey: .activeTurnStartedAt)
        activeAt = try container.decodeIfPresent(Int.self, forKey: .activeAt)
        updatedAt = try container.decodeIfPresent(Int.self, forKey: .updatedAt)
        metadata = try container.decodeIfPresent(VersionedValue<SessionMetadata>.self, forKey: .metadata)
        agentState = try container.decodeIfPresent(VersionedValue<AgentState>.self, forKey: .agentState)
        todos = try container.decodeIfPresent(VersionedValue<[TodoItem]>.self, forKey: .todos)
        teamState = try container.decodeIfPresent(VersionedValue<JSONValue>.self, forKey: .teamState)
        model = try Self.decodePatchField(String.self, in: container, forKey: .model)
        modelReasoningEffort = try Self.decodePatchField(String.self, in: container, forKey: .modelReasoningEffort)
        effort = try Self.decodePatchField(String.self, in: container, forKey: .effort)
        serviceTier = try Self.decodePatchField(String.self, in: container, forKey: .serviceTier)
        permissionMode = try container.decodeIfPresent(PermissionMode.self, forKey: .permissionMode)
        collaborationMode = try container.decodeIfPresent(CodexCollaborationMode.self, forKey: .collaborationMode)
        copilotAgentMode = try container.decodeIfPresent(CopilotAgentMode.self, forKey: .copilotAgentMode)
        backgroundTaskCount = try container.decodeIfPresent(Int.self, forKey: .backgroundTaskCount)
        scratchlistUpdatedAt = try container.decodeIfPresent(Int.self, forKey: .scratchlistUpdatedAt)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(active, forKey: .active)
        try container.encodeIfPresent(thinking, forKey: .thinking)
        try Self.encodePatchField(activeTurnStartedAt, in: &container, forKey: .activeTurnStartedAt)
        try container.encodeIfPresent(activeAt, forKey: .activeAt)
        try container.encodeIfPresent(updatedAt, forKey: .updatedAt)
        try container.encodeIfPresent(metadata, forKey: .metadata)
        try container.encodeIfPresent(agentState, forKey: .agentState)
        try container.encodeIfPresent(todos, forKey: .todos)
        try container.encodeIfPresent(teamState, forKey: .teamState)
        try Self.encodePatchField(model, in: &container, forKey: .model)
        try Self.encodePatchField(modelReasoningEffort, in: &container, forKey: .modelReasoningEffort)
        try Self.encodePatchField(effort, in: &container, forKey: .effort)
        try Self.encodePatchField(serviceTier, in: &container, forKey: .serviceTier)
        try container.encodeIfPresent(permissionMode, forKey: .permissionMode)
        try container.encodeIfPresent(collaborationMode, forKey: .collaborationMode)
        try container.encodeIfPresent(copilotAgentMode, forKey: .copilotAgentMode)
        try container.encodeIfPresent(backgroundTaskCount, forKey: .backgroundTaskCount)
        try container.encodeIfPresent(scratchlistUpdatedAt, forKey: .scratchlistUpdatedAt)
    }

    private static func decodePatchField<Value: Codable & Equatable & Sendable>(
        _ type: Value.Type,
        in container: KeyedDecodingContainer<CodingKeys>,
        forKey key: CodingKeys
    ) throws -> PatchField<Value>? {
        guard container.contains(key) else { return nil }
        if try container.decodeNil(forKey: key) { return .null }
        return .value(try container.decode(Value.self, forKey: key))
    }

    private static func encodePatchField<Value: Codable & Equatable & Sendable>(
        _ field: PatchField<Value>?,
        in container: inout KeyedEncodingContainer<CodingKeys>,
        forKey key: CodingKeys
    ) throws {
        guard let field else { return }
        switch field {
        case .null:
            try container.encodeNil(forKey: key)
        case .value(let value):
            try container.encode(value, forKey: key)
        }
    }
}

File: android/core/protocol/src/main/kotlin/app/hapi/protocol/wire/SessionPatch.kt
package app.hapi.protocol.wire

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject

/**
 * Atomic `{version, value}` pair used by structured `session-updated` patches
 * for `metadata` / `agentState` / `todos` / `teamState`. The version is the
 * only safe way to reject stale patches — the dual SSE connections have no
 * shared ordering. Never assign the wrapper itself into a [Session]; apply
 * `value` and store `version` (see `docs/api/client-contract/sse.md`).
 */
data class VersionedValue<T>(
    /**
     * Write counter (metadata/agentState) or epoch-ms watermark
     * (todos/teamState) — compared strictly-greater against the cached
     * watermark on [Session].
     */
    val version: Long,
    val value: T?,
)

/**
 * Decoded `SessionPatch` (`SessionPatchSchema`, `shared/src/schemas.ts`
 * — strict). Build via [SessionPatches.parse]; a data-class instance is also
 * fine for tests/local use.
 *
 * Field encoding of zod optionality:
 * - optional non-nullable fields → plain `T?` where `null` means absent;
 * - `nullable().optional()` fields ([activeTurnStartedAt], [model],
 *   [modelReasoningEffort], [effort], [serviceTier]) → [OptionalField], because
 *   an explicit wire `null` clears the target field while absence leaves it
 *   untouched.
 *
 * Note `activeTurnStartedAt` and `scratchlistUpdatedAt` are carried but never
 * assigned by `applySessionDetailPatch` — replicating the reference client
 * (see `app.hapi.protocol.patch.SessionPatching`).
 */
data class SessionPatch(
    val active: Boolean? = null,
    val thinking: Boolean? = null,
    val activeTurnStartedAt: OptionalField<Long?> = OptionalField.Absent,
    val activeAt: Long? = null,
    val updatedAt: Long? = null,
    val metadata: VersionedValue<SessionMetadata>? = null,
    val agentState: VersionedValue<AgentState>? = null,
    val todos: VersionedValue<List<TodoItem>>? = null,
    val teamState: VersionedValue<JsonElement>? = null,
    val model: OptionalField<String?> = OptionalField.Absent,
    val modelReasoningEffort: OptionalField<String?> = OptionalField.Absent,
    val effort: OptionalField<String?> = OptionalField.Absent,
    val serviceTier: OptionalField<String?> = OptionalField.Absent,
    val permissionMode: String? = null,
    val collaborationMode: String? = null,
    val copilotAgentMode: String? = null,
    val backgroundTaskCount: Int? = null,
    /** Bare refetch trigger for `GET /sessions/:id/scratchlist`; carries no data. */
    val scratchlistUpdatedAt: Long? = null,
)

object SessionPatches {
    private val KNOWN_KEYS = setOf(
        "active", "thinking", "activeTurnStartedAt", "activeAt", "updatedAt",
        "metadata", "agentState", "todos", "teamState",
        "model", "modelReasoningEffort", "effort", "serviceTier",
        "permissionMode", "collaborationMode", "copilotAgentMode",
        "backgroundTaskCount", "scratchlistUpdatedAt",
    )

    /**
     * Parse a `session-updated` data payload as a [SessionPatch], mirroring
     * the web reference `getSessionPatch` (`SessionPatchSchema.safeParse` +
     * non-empty check): returns `null` for anything that is not a valid,
     * non-empty patch — the caller then treats the payload as a full
     * [Session] or falls back to a REST refetch.
     *
     * Strictness is load-bearing: `SessionPatchSchema` is `.strict()`, so an
     * unknown key (e.g. `id` on a full session object) must fail. Total —
     * never throws.
     *
     * Deliberate leniencies vs zod (noted, not accidental): enum-typed
     * strings (`permissionMode` etc.) accept any string so a future hub mode
     * cannot force a refetch loop, and `teamState.value` is only shallowly
     * checked (object with string `teamName`, or null).
     */
    fun parse(element: JsonElement?): SessionPatch? {
        val obj = element as? JsonObject ?: return null
        if (obj.isEmpty() || obj.keys.any { it !in KNOWN_KEYS }) return null
        return try {
            SessionPatch(
                active = obj["active"]?.let { it.boolOrNull ?: return null },
                thinking = obj["thinking"]?.let { it.boolOrNull ?: return null },
                activeTurnStartedAt = optionalNullableLong(obj, "activeTurnStartedAt") ?: return null,
                activeAt = obj["activeAt"]?.let { it.longOrNull ?: return null },
                updatedAt = obj["updatedAt"]?.let { it.longOrNull ?: return null },
                metadata = obj["metadata"]?.let { wrapper ->
                    versioned(wrapper) { value ->
                        if (value is JsonNull) null
                        else HapiJson.decodeFromJsonElement(SessionMetadata.serializer(), value)
                    } ?: return null
                },
                agentState = obj["agentState"]?.let { wrapper ->
                    versioned(wrapper) { value ->
                        if (value is JsonNull) null
                        else HapiJson.decodeFromJsonElement(AgentState.serializer(), value)
                    } ?: return null
                },
                todos = obj["todos"]?.let { wrapper ->
                    versioned(wrapper) { value ->
                        if (value !is JsonArray) throw IllegalArgumentException("todos.value must be an array")
                        value.map { HapiJson.decodeFromJsonElement(TodoItem.serializer(), it) }
                    } ?: return null
                },
                teamState = obj["teamState"]?.let { wrapper ->
                    versioned(wrapper) { value ->
                        when {
                            value is JsonNull -> null // TeamDelete clear
                            value is JsonObject && value["teamName"].stringOrNull != null -> value
                            else -> throw IllegalArgumentException("teamState.value must be null or a team object")
                        }
                    } ?: return null
                },
                model = optionalNullableString(obj, "model") ?: return null,
                modelReasoningEffort = optionalNullableString(obj, "modelReasoningEffort") ?: return null,
                effort = optionalNullableString(obj, "effort") ?: return null,
                serviceTier = optionalNullableString(obj, "serviceTier") ?: return null,
                permissionMode = obj["permissionMode"]?.let { it.stringOrNull ?: return null },
                collaborationMode = obj["collaborationMode"]?.let { it.stringOrNull ?: return null },
                copilotAgentMode = obj["copilotAgentMode"]?.let { raw ->
                    val mode = raw.stringOrNull ?: return null
                    // zod coerces the legacy `'fleet'` literal to 'interactive'.
                    if (mode == "fleet") "interactive" else mode
                },
                backgroundTaskCount = obj["backgroundTaskCount"]?.let { it.intOrNull ?: return null },
                scratchlistUpdatedAt = obj["scratchlistUpdatedAt"]?.let { it.longOrNull ?: return null },
            )
        } catch (_: Exception) {
            null
        }
    }

    /** `{version, value}` wrapper: `version` must be a number, `value` key must exist. */
    private fun <T> versioned(wrapper: JsonElement, decodeValue: (JsonElement) -> T?): VersionedValue<T>? {
        val obj = wrapper as? JsonObject ?: return null
        val version = obj["version"].longOrNull ?: return null
        if (!obj.containsKey("value")) return null
        return try {
            VersionedValue(version, decodeValue(obj.getValue("value")))
        } catch (_: Exception) {
            null
        }
    }

    /**
     * Tri-state for `z.number().nullable().optional()`: absent → Absent,
     * `null` → Present(null), number → Present(n); anything else → null (invalid).
     */
    private fun optionalNullableLong(obj: JsonObject, key: String): OptionalField<Long?>? {
        val raw = obj[key] ?: return OptionalField.Absent
        if (raw is JsonNull) return OptionalField.Present(null)
        return raw.longOrNull?.let { OptionalField.Present(it) }
    }

    /** Tri-state for `z.string().nullable().optional()`. */
    private fun optionalNullableString(obj: JsonObject, key: String): OptionalField<String?>? {
        val raw = obj[key] ?: return OptionalField.Absent
        if (raw is JsonNull) return OptionalField.Present(null)
        return raw.stringOrNull?.let { OptionalField.Present(it) }
    }
}
